content/posts/cicd_ha.md

```diff
@@ -0,0 +1,132 @@
+
+
+## What is HA and why is it relative?
+
+Even though [the definition at Wikipedia](https://en.wikipedia.org/wiki/High_availability) is not a true "official" source, I like how it defines HA:
+
+> *High availability (HA) is a characteristic of a system, which aims to ensure an agreed level of operational performance, usually uptime, for a higher than normal period.*
+
+The words *agreed level of operational performance* show some very important topics to clarify:
+
+* *Operational performance* is defined by the use case and the business. Not from the technical components or the software being used.
+* The software architecture and its technical definition depends on the *agreed level* of that operational performance.
+
+Let's use an example. If we think about the service that [Amazon retail](https://amazon.com) users are expecting, any purchase transaction using the web portal cannot be interrupted for even a few seconds. So, web services, front end applications, databases, middleware, payment channels  and other services or microservices, etc. cannot interrupt or block that user transaction. That means that every server needs to be up and running all the time, or at least to recover at the same point in a sub-second basis. 
+
+This is just because the business requires that any user transaction and interaction with the application cannot be interrupted at all, considering that an interruption is something blocked during 3 seconds or more (users don't like to wait more than 3 seconds to refresh a web page to the next step). And if that doesn't happen, I - as a user - will be doing the transaction from a different web retailer or competitor. Business then requires a very restrictive HA scenario.
+
+I am not going to enter in the differences of *"Active-Active"* or *"Active-Passive"* HA scenarios. But in the previous example a software architecture candidate usually needs to be based on an "Active-Active" replication method, where every software transaction is supported or backed-up automatically from an already running service that can re-take immediately the current step in the transaction, with no data-loss, or acting as a performant load balancer solution.
+
+But not every use case or business requires the same level of HA. In other terms, the word *High* in the *HA* term is relative. A critical service interruption can be seconds, minutes or even hours depending on the context, use case or business.
+
+## What about HA and CI/CD?
+
+Continuous Integration and Continuous Delivery (CI/CD) is based on the concept of continuously develop, build, integrate, test, deliver and release the software, with the end goal of releasing high quality applications for demanding users. So, how should we define a highly available platform or solution to be able to do that with no *"service interruption"*?
```

Comment:
> continuously develop, build, integrate, test, deliver and release the software

Develop, build, etc. should be replaced with developing, building, testing, etc.